File: basic_dsa_one_hundred_eighty_nine.java
/**
 * Basic DSA Problems - Remove Nth Node From End of List
 * Remove the nth node from the end of a linked list (one pass)
 */
public class basic_dsa_one_hundred_eighty_nine {

    // Problem: Given head and n, remove the nth node from the end. Return head.

    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }

    /**
     * Two pointers: advance first by n+1 steps, then move both until first is null; second.next is to remove.
     * Time O(L), Space O(1)
     */
    public static ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy, slow = dummy;
        for (int i = 0; i <= n; i++) fast = fast.next;
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        slow.next = slow.next.next;
        return dummy.next;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Remove Nth Node From End of List");
        System.out.println("=".repeat(60));

        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);
        ListNode result = removeNthFromEnd(head, 2);
        StringBuilder sb = new StringBuilder();
        while (result != null) {
            sb.append(result.val).append(" ");
            result = result.next;
        }
        System.out.println("After removing 2nd from end -> " + sb.toString().trim());
    }
}

File: basic_dsa_one_hundred_eighty_nine.py
"""
Basic DSA - Pascal's Triangle II
Given an integer row_index, return the row_index-th (0-indexed) row of Pascal's triangle.
"""


def get_row(row_index):
    """
    Build the row in-place from right to left.
    Time Complexity: O(row_index^2)
    Space Complexity: O(row_index)
    """
    row = [0] * (row_index + 1)
    row[0] = 1
    for i in range(1, row_index + 1):
        for j in range(i, 0, -1):
            row[j] += row[j - 1]
    return row


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Pascal's Triangle II")
    print("=" * 60)

    print(get_row(0))  # [1]
    print(get_row(3))  # [1, 3, 3, 1]
